includes/Checker/Checks/Trademarks_Check.php

```diff
@@ -0,0 +1,482 @@
+
+		// Filter the readme files located at root.
+		$potential_readme_files = array_filter(
+			$readme_list,
+			function ( $file ) use ( $plugin_relative_path ) {
+				$file = str_replace( $plugin_relative_path, '', $file );
+				if ( ! strpos( $file, '/' ) ) {
+					return true;
+				}
+			}
+		);
+
+		// Find the .txt versions of the readme files.
+		$readme_txt = array_filter(
+			$potential_readme_files,
+			function ( $file ) {
+				return preg_match( '/^readme\.txt$/i', basename( $file ) );
+			}
+		);
+
+		// Find the .md versions of the readme files.
+		$readme_md = array_filter(
+			$potential_readme_files,
+			function ( $file ) {
+				return preg_match( '/^readme\.md$/i', basename( $file ) );
+			}
+		);
+
+		// If there's a .txt version, ignore .md versions.
+		$readme = ( ! empty( $readme_txt ) ) ? $readme_txt : $readme_md;
+
+		// If the readme file does not exist, then skip test.
+		if ( empty( $readme ) ) {
+			return;
+		}
+
+		$matches = array();
+		// Get the plugin name from readme file.
+		$file = self::file_preg_match( '/===(.*)===/i', $files, $matches );
+
+		if ( ! $file ) {
+			return;
+		}
+
+		$name = isset( $matches[1] ) ? $matches[1] : '';
+
+		$this->look_for_trademark(
+			$result,
+			str_replace( $result->plugin()->path(), '', $file ),
+			$name,
+			__( 'The plugin name includes a restricted term.', 'plugin-check' )
+		);
+	}
+
+	/**
+	 * Checks the readme file for default text.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The Check Result to amend.
+	 */
+	private function check_for_name( Check_Result $result ) {
+		if ( ! function_exists( 'get_plugin_data' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/plugin.php';
+		}
+
+		$plugin_main_file = WP_PLUGIN_DIR . '/' . $result->plugin()->basename();
+		$plugin_header    = get_plugin_data( $plugin_main_file );
+
+		if ( ! empty( $plugin_header['Name'] ) ) {
+			$this->look_for_trademark(
+				$result,
+				$plugin_main_file,
+				$plugin_header['Name'],
+				__( 'The plugin name includes a restricted term.', 'plugin-check' )
+			);
+		}
+	}
+
+	/**
+	 * Checks the readme file for default text.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The Check Result to amend.
+	 */
+	private function check_for_slug( Check_Result $result ) {
+		$plugin_slug = basename( $result->plugin()->path() );
+
+		$this->look_for_trademark(
+			$result,
+			WP_PLUGIN_DIR . '/' . $result->plugin()->basename(),
+			$plugin_slug,
+			__( 'The plugin slug includes a restricted term.', 'plugin-check' )
+		);
+	}
+
+	/**
+	 * Determines if we find a trademarked term.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result   The Check Result to amend.
+	 * @param string       $file     The plugin file.
+	 * @param string       $input    The plugin name or slug.
+	 * @param string       $preamble The preamble message.
+	 */
+	private function look_for_trademark( $result, $file, $input, $preamble ) {
```

Comment:
This method does too many things at once and therefore becomes overly complicated. A few problems I see in this method:
* I think we should simplify the `$input` parameter. It is confusing that it can be either a slug or a name, especially because this will lead to confusing error messages below. If you pass a slug for `$input`, the messages will still say "Your plugin name - ...", although it is actually a plugin slug. We should probably rename this parameter to `$plugin_name` and the method to `look_for_trademarked_name()`. For the slug check, we should then use another method like `look_for_trademarked_slug()`, and move the common parts into lower-level methods that both can share.
* The `$preamble` parameter isn't necessary. As far as I can tell, the message for it is always the same, so we should just put it directly into the function.
* The messages should probably also be different for the slug specific check so that they are not confusing about what the problem is.

Maybe you can further simplify this by not requiring the `$result` and `$file` parameters even and simply throwing exceptions for any trademark problems. For example as follows:
* `validate_name_has_no_trademarks( $plugin_name )` (throws exceptions with the specific messages for name problems)
* `validate_slug_has_no_trademarks( $plugin_slug )` (throws exceptions with the specific messages for slug problems)
* Implement the `add_result_error_for_file()` method here that other checks have too.
* In the other methods, you can then use something like:
```php
try {
	$this->validate_name_has_no_trademarks( $plugin_name );
	// Or use $this->validate_slug_has_no_trademarks( $plugin_slug );
} catch ( Exception $e ) {
	$this->add_result_error_for_file( $result, $file, 'trademarked_term', $e->getMessage() );
}
```

This is so little logic that it isn't really duplicate code.

Reply:
Thank you for the valuable feedback. I have updated the code in accordance with your guidance, which can be found at https://github.com/WordPress/plugin-check/pull/285/commits/c43ee3cdf220dbf7c769ac1175095b5048e3e31f.